ultravox/training/evaluation.py

```diff
@@ -156,21 +160,23 @@ def evaluate(
 
         scores = [x for x in possibly_non_scores if x is not None]
 
-        if verbose:
-            print(f"Eval for {task.dataset}:")
-            for sample, score in zip(output_samples, scores):
-                print("-" * 20)
-                print(f"Q: {sample.question}")
-                print(f"A: {sample.generated_answer}")
-                print(f"X: {sample.expected_answer} [score: {score:.2f}]")
-
         average = np.mean(scores)
         std = np.std(scores) / np.sqrt(len(scores))
         metrics[f"eval_{task.name}"] = average
         metrics[f"eval_{task.name}_std"] = std
 
-        print(
-            f"Aggregate {task.metric} score for {task.dataset}: {average:.2f} ± {std:.2f}"
-        )
+        has_audio_str = "with" if task.include_audio else "without"
+        agg_score_str = f"Aggregate {task.metric} score for {task.dataset} ({has_audio_str} audio): {average:.2f} ± {std:.2f}"
+        print(agg_score_str)
+
+        if log_dir:
+            with open(os.path.join(log_dir, f"{task.name}.json"), "w") as f:
+                f.write(f"Task info: {str(task)}\n")
+                f.write(agg_score_str + "\n")
+                for sample, score in zip(output_samples, scores):
+                    f.write("-" * 20 + "\n")
+                    f.write(f"Q: {sample.question}\n")
+                    f.write(f"A: {sample.generated_answer}\n")
+                    f.write(f"X: {sample.expected_answer} [score: {score:.2f}]\n")
```

Comment:
For logging the results, how does this compare to simply printing in jsonl format with the following keys: Q, A, X, S or question, answer, expected, score? Using jsonl format makes it easier if we need to parse the result.

Reply:
The focus was on keeping human readability and matching existing formats.

I thought about JSONL, but having all of this info in one line is gonna make it unreadable.
With JSON we might be able to get something, but still we'd have to compromise readability a bit.

The question is: how important is it to read these programatically vs by humans.
We already have a snippet that can parse this, but if the programmatic readability becomes more important and we want to do it easily multiple times then we can do JSON.

Reply:
We can easily use tools like jq to pretty-print jsonl files on demand.

Reply:
Wait, by "on demand" do you mean you want to do this in the command-line?
The point of logging is to make it as easy as possible to use. Downloading, then using `jq` seems like two steps too many for just seeing the output of evals.

I don't know if W&B gives you a CLI for interacting with Files. I see a Python API, but that's not gonna be convenient without extra tooling support.

Reply:
As long as we can manually download the files, it is okay. We will not be checking eval outputs for every experiment run. Our focus will mainly be on the aggregated scores that are already there in w&b. The eval output logging will primarily be used for analysis.

Reply:
I disagree. Outputs should be easily inspectable.

Remember [this Slack thread](https://fixie-ai.slack.com/archives/C06CXPY070V/p1722032451121769?thread_ts=1722032121.134989&cid=C06CXPY070V) and the exploration to compare sentence-bleu vs corpus-bleu?

That whole exploration need not have occurred if the actual outputs were glanced at first.

Reply:
Anyway, I'll do JSON with `indent=1`.

Reply:
I agree that human readability is worth preserving, JSONLs are notoriously hard to eyeball.

Reply:
That's what I meant by analysis. We downloaded the output files, parsed the results, computed and verified scores, and then realized that the problem was with prompting. 

We want make it easier to perform analysis, either programmatically or by human, and it goes beyond functionalities provided on wandb.

Reply:
If we wan to have a balance, maybe choose json instead of jsonl.

Reply:
> That's what I meant by analysis. We downloaded the output files, parsed the results, computed and verified scores, and then realized that the problem was with prompting.

Again, going back to the example (not trying to make this personal, just using this as example): you didn't do it on the first time, which proves my point: people do what's easiest. If you knew you could just look at logs and they were easy to check, you would've done that first.

Reply:
It's JSON now.

Reply:
You are right. Actually I didn’t catch the problem when I first looked at the downloaded files. I reran bleu calculation, verified your results, and then started to look for other causes. Anything we do to make it easier to spot errors would be helpful.